### PR TITLE
clean doesn't try to remove older entries on 1 elem slot list

### PR DESCRIPTION
**Pull request by jeffwashington**
#### Problem
clean tries to remove older entries in the slot list for entries that only have 1 element in the slot list. This is inefficient.

#### Summary of Changes
Only try to remove older slot list entries when there can be older slot list entries.

Fixes #

### CODE REVIEW

File: accounts-db/src/accounts_index.rs [hunk 1]
```diff
@@ -1817,6 +1820,10 @@ impl<T: IndexValue, U: DiskIndexValue + From<T> + Into<T>> AccountsIndex<T, U> {
         reclaims: &mut SlotList<T>,
         max_clean_root_inclusive: Option<Slot>,
     ) {
+        if slot_list.len() <= 1 {
+            self.purge_older_root_entries_one_slot_list
```
Comment (jeffwashington) [inline accounts-db/src/accounts_index.rs:1824]: this has a high cost and should be avoided. We are marking the index entry as dirty no matter what below here. if this ever occurs again, we'd like to know.

File: accounts-db/src/accounts_db.rs [hunk 2]
```diff
@@ -3284,7 +3285,14 @@ impl AccountsDb {
                                                 {
                                                     assert!(slot <= &max_clean_root_inclusive);
                                                 }
-                                                purges_old_accounts.push(*candidate);
+                                                if slot_list.len() > 1 {
+                                                    // no need to purge old accounts if there is only 1 slot in the slot list
+                                                    purges_old_accounts.push(*candidate);
```
Comment (jeffwashington) [inline accounts-db/src/accounts_db.rs:3290]: the key observation here is that when `purges_old_accounts` is used, it only does useful work when the slot_list.len() > 1:
https://github.com/anza-xyz/agave/blob/d7353b038be622ac265b90536dda071977a9d926/accounts-db/src/accounts_db.rs#L2680

**Review by HaoranYi — DISMISSED** [review]
lgtm!

File: accounts-db/src/accounts_db.rs [hunk 3]
```diff
@@ -3284,7 +3285,14 @@ impl AccountsDb {
                                                 {
                                                     assert!(slot <= &max_clean_root_inclusive);
                                                 }
-                                                purges_old_accounts.push(*candidate);
+                                                if slot_list.len() > 1 {
```
Comment (brooksprumo) [inline accounts-db/src/accounts_db.rs:3288]: What happens if the slot list is empty?

File: accounts-db/src/accounts_db.rs [hunk 4]
```diff
@@ -3284,7 +3285,14 @@ impl AccountsDb {
...
+                                                    // no need to purge old accounts if there is only 1 slot in the slot list
+                                                    purges_old_accounts.push(*candidate);
+                                                } else {
+                                                    self.clean_accounts_stats
+                                                        .uncleaned_roots_slot_list_1
+                                                        .fetch_add(1, Ordering::Relaxed);
+                                                }
```
Comment (brooksprumo) [inline accounts-db/src/accounts_db.rs:3296]: Should `useless` be `true` in this `else`? (Or really, move the `useless = false;` line (below) up into the `if` case (above).)

File: accounts-db/src/accounts_db.rs (hunk 3 above)
Comment (jeffwashington) [reply accounts-db/src/accounts_db.rs:3288]: slot_list.len() cannot be 0 here. We are in a match arm:
`Some(index_in_slot_list)` indicating we found an alive entry in the slot list at index: `index_in_slot_list`.

File: accounts-db/src/accounts_db.rs (hunk 4 above)
Comment (jeffwashington) [reply accounts-db/src/accounts_db.rs:3296]: i moved `useless`

**Review by brooksprumo — COMMENTED** [review]
:shipit:

**Review by brooksprumo — APPROVED** [review]

**Review by HaoranYi — APPROVED** [review]

**Pull request merged**